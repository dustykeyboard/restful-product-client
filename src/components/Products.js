import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import qs from 'query-string'
import Product from './Product'
import '../styles/products.css'

class Products extends PureComponent {
  constructor(props) {
    super(props)

    this.state = {
      page: '1',
      perpage: '12',
      products: [],
      pages: 1,
    }

    this.handlePageChange = this.handlePageChange.bind(this)
    this.handlePerPageChange = this.handlePerPageChange.bind(this)
    this.renderControls = this.renderControls.bind(this)
  }

  static getDerivedStateFromProps(props, state) {
    const perpage = parseInt(state.perpage, 10)
    const page = parseInt(state.page, 10)

    const pages = Math.ceil(props.products.items.length / perpage)
    const begin = Math.max(perpage * (page - 1), 0)
    const end = begin + parseInt(state.perpage, 10)

    return {
      products: props.products.items.slice(begin, end),
      pages,
      page: Math.min(page, pages),
    }
  }

  componentDidMount() {
    const { loaded, loading, fetchProducts } = this.props
    if (!loaded && !loading) fetchProducts()

    const queryString = this.props.location && this.props.location.search
    const { page = 1, perpage = 12 } = qs.parse(queryString)

    this.setState({ page, perpage })
  }

  handlePageChange(event) {
    const page = event.target.value
    this.props.updateParams({ ...this.state, page })
    this.setState({ page })
  }

  handlePerPageChange(event) {
    const perpage = event.target.value
    this.props.updateParams({ ...this.state, perpage })
    this.setState({ perpage })
  }

  renderControls() {
    const { products: { count } } = this.props
    const { pages, page, perpage } = this.state

    const pageOptions = []
    for (var i = 1; i <= pages; i++) {
      pageOptions.push(i)
    }
    const perPageOptions = [12, 30, 60, 120]

    return (
      <div className="products__toolbar">
        <div className="product__count">{count} Products</div>
        <div className="products__controls">
          <div className="products__perpage">
            <select onChange={this.handlePerPageChange} value={perpage}>
              {perPageOptions.map(o => (
                <option key={o} value={o}>
                  {o} per page
                </option>
              ))}
            </select>
          </div>

          <div className="products__pagination">
            Page&nbsp;
            <select onChange={this.handlePageChange} value={page}>
              {pageOptions.map(o => (
                <option key={o} value={o}>
                  {o}
                </option>
              ))}
            </select>
          </div>
        </div>
      </div>
    )
  }

  render() {
    const { products: { loading, error } } = this.props
    const { products } = this.state

    return (
      <div className="page page__products">
        <h2>All Products</h2>

        {error && <p className="error">{error.message}</p>}

        {loading && <div className="loading">Loading Products</div>}

        {products.length > 0 && (
          <div>
            {this.renderControls()}

            <div className="products">
              {products.map(product => (
                <Product key={product.id} {...product} />
              ))}
            </div>

            {this.renderControls()}
          </div>
        )}
      </div>
    )
  }
}

Products.propTypes = {
  products: PropTypes.shape({
    loading: PropTypes.bool,
    error: PropTypes.oneOf([
      false,
      PropTypes.shape({
        message: PropTypes.string,
      }),
    ]),
    count: PropTypes.number,
    items: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        product_name: PropTypes.string,
      }),
    ),
  }),
  fetchProducts: PropTypes.func.isRequired,
  updateParams: PropTypes.func.isRequired,
}

export default Products
