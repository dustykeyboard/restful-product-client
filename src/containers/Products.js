import { connect } from 'react-redux'
import { push } from 'connected-react-router'
import qs from 'query-string'
import { actions, selector } from '../models/products'
import Products from '../components/Products'

const mapStateToProps = (state, props) => ({
  products: selector(state),
})

const mapDispatchToProps = dispatch => ({
  fetchProducts: () => dispatch(actions.request()),
  updateParams: ({ page, perpage }) => {
    dispatch(push({ search: qs.stringify({ page, perpage }) }))
  },
})

export default connect(mapStateToProps, mapDispatchToProps)(Products)
