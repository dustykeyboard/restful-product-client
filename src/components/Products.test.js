import React from 'react'
import renderer from 'react-test-renderer'
import Products from './Products'

const mockProps = overrides => ({
  products: {
    loading: true,
    loaded: true,
    error: false,
    items: [
      {
        id: 1,
        product_name: 'product',
        description: 'description',
        product_image: 'product.png',
        price: '$0.99',
      },
    ],
    count: 0,
  },
  fetchProducts: jest.fn(),
  updateParams: jest.fn(),
  ...overrides,
})

describe('Products component', () => {
  it('renders', () => {
    const renderedComponent = renderer
      .create(<Products {...mockProps()} />)
      .toJSON()

    expect(renderedComponent).toMatchSnapshot()
  })

  it('requests products', () => {
    const fetchMock = jest.fn()
    const renderedComponent = renderer
      .create(<Products {...mockProps({ fetchProducts: fetchMock })} />)
      .toJSON()

    expect(fetchMock.mock.calls.length).toBe(1)
  })
})
