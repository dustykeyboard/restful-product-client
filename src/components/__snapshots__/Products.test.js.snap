// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products component renders 1`] = `
<div>
  <h2>
    All Products
  </h2>
  <div
    className="loading"
  >
    Loading Products
  </div>
  <div>
    <div
      className="controls"
    >
      <div
        className="productCount"
      >
        0
         Products
      </div>
      <div
        className="perPage"
      >
        <select
          onChange={[Function]}
          value={12}
        >
          <option
            value={12}
          >
            12
             per page
          </option>
          <option
            value={30}
          >
            30
             per page
          </option>
          <option
            value={60}
          >
            60
             per page
          </option>
          <option
            value={120}
          >
            120
             per page
          </option>
        </select>
      </div>
      <div
        className="pagination"
      >
        Page
        <select
          onChange={[Function]}
          value={1}
        >
          <option
            value={1}
          >
            1
          </option>
        </select>
      </div>
    </div>
    <div
      className="products"
    >
      <div
        className="product"
      >
        <img
          alt="{product_name}"
          className="product_image"
          height={340}
          src="product.png"
          width={340}
        />
        <h2
          className="product_name"
        >
          product
        </h2>
        <p
          className="product_description"
        >
          description
        </p>
        <p
          className="product_price"
        >
          $0.99
        </p>
      </div>
    </div>
    <div
      className="controls"
    >
      <div
        className="productCount"
      >
        0
         Products
      </div>
      <div
        className="perPage"
      >
        <select
          onChange={[Function]}
          value={12}
        >
          <option
            value={12}
          >
            12
             per page
          </option>
          <option
            value={30}
          >
            30
             per page
          </option>
          <option
            value={60}
          >
            60
             per page
          </option>
          <option
            value={120}
          >
            120
             per page
          </option>
        </select>
      </div>
      <div
        className="pagination"
      >
        Page
        <select
          onChange={[Function]}
          value={1}
        >
          <option
            value={1}
          >
            1
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;
